connectors/filtering/basic_rule.py

```diff
@@ -0,0 +1,177 @@
+# you may not use this file except in compliance with the Elastic License 2.0.
+#
+import datetime
+import re
+from enum import Enum
+
+from dateutil.parser import ParserError, parser
+
+from connectors.logger import logger
+
+
+class BasicRule:
+    DEFAULT_RULE_ID = "DEFAULT"
+
+    class Policy(Enum):
+        INCLUDE = 1
+        EXCLUDE = 2
+
+    class Rule(Enum):
+        EQUALS = 1
+        STARTS_WITH = 2
+        ENDS_WITH = 3
+        CONTAINS = 4
+        REGEX = 5
+        GREATER_THAN = 6
+        LESS_THAN = 7
+
+    def __init__(self, id_, order, policy, field, rule, value):
+        self._id = id_
+        self._order = order
+        self._policy = policy
+        self._field = field
+        self._rule = rule
+        self._value = value
+
+    @staticmethod
+    def default_rule():
+        return BasicRule(
+            id_=BasicRule.DEFAULT_RULE_ID,
+            order=0,
+            policy=BasicRule.Policy.INCLUDE,
+            field="_",
+            rule=BasicRule.Rule.EQUALS,
+            value=".*",
+        )
+
+    def matches_document(self, document):
+        if self.is_default_rule:
+            return True
+
+        if self.field not in document:
+            return False
+
+        document_value = document[self.field]
+        coerced_rule_value = self.coerce_rule_value_based_on_document_value(
+            document_value
+        )
+
+        match self.rule:
+            case BasicRule.Rule.STARTS_WITH:
+                return str(document_value).startswith(self.value)
+            case BasicRule.Rule.ENDS_WITH:
+                return str(document_value).endswith(self.value)
+            case BasicRule.Rule.CONTAINS:
+                return self.value in str(document_value)
+            case BasicRule.Rule.REGEX:
+                return re.match(self.value, str(document_value)) is not None
+            case BasicRule.Rule.LESS_THAN:
+                return document_value < coerced_rule_value
+            case BasicRule.Rule.GREATER_THAN:
+                return document_value > coerced_rule_value
+            case BasicRule.Rule.EQUALS:
+                return document_value == coerced_rule_value
+
+        return False
+
+    @property
+    def id_(self):
+        return self._id
+
+    @property
+    def order(self):
+        return self._order
+
+    @property
+    def policy(self):
+        return self._policy
+
+    @property
+    def field(self):
+        return self._field
+
+    @property
+    def rule(self):
+        return self._rule
+
+    @property
+    def value(self):
+        return self._value
+
+    @property
+    def is_default_rule(self):
+        return self.id_ == BasicRule.DEFAULT_RULE_ID
+
+    @property
+    def is_include(self):
+        return self.policy == BasicRule.Policy.INCLUDE
+
+    def coerce_rule_value_based_on_document_value(self, doc_value):
+        try:
+            match doc_value:
+                case str():
```

Comment:
Wouldn't match case be more pythonic with [PEP 634](https://peps.python.org/pep-0634/#class-patterns)?

Reply:
I have never used that pattern before because each case creates an instance of an object just for comparison, and this is a very new syntax -- a bit of uncharted territory. 

So I don't know :) 

We can do this; it does look nice. In general, we try to avoid extra instance creations (not pythonic), and this is my first reaction when reviewing.